File: 39-combination-sum/combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> comb;
        backtrack(0,0,candidates,target,res,comb);
        return res;
    }
    void backtrack(int a,int sum,vector<int>& candidates,int target,vector<vector<int>>& res,vector<int> comb){
            if(sum==target){
                res.push_back(comb);
                return ;
            }
            if(sum>target || a>=candidates.size()){
                return ;
            }
            comb.push_back(candidates[a]);
            backtrack(a,sum+candidates[a],candidates,target,res,comb);
            comb.pop_back();
            backtrack(a+1,sum,candidates,target,res,comb);


        
    }

};

File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> ans=new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        backtrack(0,candidates,target,new ArrayList<>());
        return ans;
    }
    private void backtrack(int begin,int []arr,int target,List<Integer> res){
        if (target == 0) {
            ans.add(new ArrayList<>(res));
            return;
        }
        if(target<0)return;
        for(int i=begin;i<arr.length;i++){
            res.add(arr[i]);
            backtrack(i,arr,target-arr[i],res);
            res.remove(res.size()-1);
        }
    }
}